### DISCUSSION TITLE
Update data in real time

### DISCUSSION THREAD

**Discussion by abaawe** [community/community#118881] [Programming Help]
### Select Topic Area

General

### Body

I am looking for a way to send a set of information to a database and receive this data again... in real time, that is, in real time.... without any delay............. I do not want to use Node js, I don't want to use RealTime Databases, I don't want to use MoongoDB, just PHP, SQL, and JavaScript.

**Comment by ketan270** [comment]
If you're looking to achieve real-time communication between a database and a web application without using Node.js, real-time databases, or MongoDB, you can consider implementing WebSockets along with PHP, SQL, and JavaScript. WebSockets enable bidirectional communication between a client (e.g., a web browser) and a server, allowing for real-time data exchange without the overhead of traditional HTTP requests.

Here's a basic outline of how you can achieve this:

Set up your database: Ensure you have a database set up with the necessary tables and columns to store the information you want to send and receive.

Create a WebSocket server in PHP: Use a library like Ratchet (http://socketo.me/) to create a WebSocket server in PHP. This server will handle WebSocket connections from clients and manage real-time data exchange.

php
Copy code
// server.php (WebSocket server using Ratchet)
require __DIR__ . '/vendor/autoload.php';

use Ratchet\MessageComponentInterface;
use Ratchet\ConnectionInterface;

class MyWebSocketServer implements MessageComponentInterface {
    protected $clients;

    public function __construct() {
        $this->clients = new \SplObjectStorage;
    }

    public function onOpen(ConnectionInterface $conn) {
        $this->clients->attach($conn);
        echo "New connection! ({$conn->resourceId})\n";
    }

    public function onMessage(ConnectionInterface $from, $msg) {
        // Handle incoming messages (e.g., store data in the database)
        // This is just an example; you should implement your own logic here
        echo "Message from {$from->resourceId}: $msg\n";
        // Store $msg in the database
    }

    public function onClose(ConnectionInterface $conn) {
        $this->clients->detach($conn);
        echo "Connection {$conn->resourceId} has disconnected\n";
    }

    public function onError(ConnectionInterface $conn, \Exception $e) {
        echo "An error occurred: {$e->getMessage()}\n";
        $conn->close();
    }
}

// Run the WebSocket server
$server = new \Ratchet\WebSocket\WsServer(new MyWebSocketServer);
$loop = \React\EventLoop\Factory::create();
$socket = new \React\Socket\Server('0.0.0.0:8080', $loop);
$server->loop = $loop;
$server->listen($socket);
$loop->run();
Create a WebSocket client in JavaScript: In your HTML file, use JavaScript to connect to the WebSocket server and send/receive data in real-time.

javascript
Copy code
// client.js (WebSocket client in JavaScript)
const socket = new WebSocket('ws://localhost:8080');

socket.onopen = function() {
    console.log('WebSocket connection established');
};

socket.onmessage = function(event) {
    console.log('Message from server:', event.data);
    // Handle incoming messages from the server (e.g., update UI)
};

socket.onclose = function(event) {
    console.log('WebSocket connection closed:', event);
};

// Send data to the server
function sendDataToServer(data) {
    socket.send(data);
}
Send data from the client: Use JavaScript to send data to the WebSocket server whenever you want to store it in the database.

javascript
Copy code
// Assuming you have a form with inputs named 'name' and 'email'
const formData = new FormData(document.getElementById('dataForm'));
const jsonData = JSON.stringify(Object.fromEntries(formData.entries()));
sendDataToServer(jsonData);
Handle data on the server: In your WebSocket server (PHP), implement logic to process incoming data and store it in the database using SQL queries.

This setup allows you to establish a real-time communication channel between your web application (using JavaScript as the client) and your server (using PHP as the WebSocket server), enabling instant data exchange and updates without the need for polling or delays.

**Comment by ebndev** [comment]
Welcome to the GitHub Community, @abaawe , we're happy you're here! 

You are more likely to get a useful response if you are posting your question in the applicable category and are explicit about what your project entails--giving a few more details might help someone give you a nudge in the right direction. 

I've gone ahead and moved your discussion to Programming Help 👩‍💻 for you. 

You might also find [this article](https://www.cloudways.com/blog/connect-mysql-with-php/#MySQL-on-Localhost) on "How to Connect a MySQL Database to PHP" helpful.

Good luck!
